Add an example of a block on a rotating disc to the tutorial documentation.

eh111eh opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

#26929

#### Brief description of what is fixed or changed
This PR is to add a system of a block on a rotating disc to the tutorial documentation. In this system, the block is located at a point on the disc surface and is subjected to forces, including friction force with additional Stribeck and viscous effects.

![IMG_5928](https://github.com/user-attachments/assets/c1212ed5-5148-4321-86a1-974e456fcff4)

#### Other comments
@moorepants @tjstienstra 

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

eh111eh committed:
Add a tutorial docs for a block on a rotating disc model and its schematic diagram (svg)

- CustomPathway is incomplete.
- Need to double check the acceleration of the point P.

eh111eh committed:
Update rotating_disc_block_friction.rst

tjstienstra commented:
@moorepants when defining a friction force we have a similar problem as when computing noncontributing forces with Kane's method. We basically want two points at the same location but with different instanteneous velocities.

A first option is to simplify the problem here removing the counter friction force, the friction force that should be applied to the disc.

The second option I see is to define a custom pathway that takes two points, which are defined by the user to be at the same location, the contact point, but with different velocities (definitions which should propagate to the inertial frame). In this case the pathway would also require a frame in which in which it has to retrieve the velocities (doesn't have to be the inertial frame).

The third option would in my opinion offers the best user interface, though it is internally more complex and it will result in slightly different definition then I would choose when defining things by hand. The problem I try to work around here by applying moving the counter force is that you don't require the two points mentioned in the second option.
```python
class SlidingPathway(Pathway):
    """"
    Parameters
    ==========
    fixed_point: Point
         Point fixed w.r.t. the provided reference frame.
    moving_point: Point
         Point with a time-dependent position w.r.t. the fixed point, where the actuator forces will be applied. This point is assumed to be fixed to another body.
    frame: ReferenceFrame
         Reference frame in which the ``fixed_point`` has a zero velocity.
    """"
    def __init__(self, fixed_point, contact_point, frame):
         self.fixed_point = fixed_point
         self.contact_point = contact_point
         self.frame = frame

    @property
    def extensions_velocity(self):
        return self.moving_point.vel(self.frame).magnitude()

    def to_loads(self, force: Expr):
         # Expansile definition like the LinearPathway
         direction = self.moving_point.vel(self.frame).normalize()
         return [
              Force(self.moving_point, force*direction),
              Force(self.fixed_point, -force*direction),
              Torque(self.frame, me.cross(self.moving_point.pos_from(self.fixed_point)*-force),
         ]
```

eh111eh committed:
Set `SlidingPathway`, 3 generalised coordinates, change position/velocity of point P and Q

- But it encounters errors with Kanes method, which I think are because of an incorrect set of position and velocity for points P and Q

moorepants commented:
> we have a similar problem as when computing noncontributing forces

But for this problem, can't we just define the force as `m*g` in the direction normal to the plane?

Also, I guess there is no reason not to allow a ReferenceFrame to be passed to a pathway. That seems fine to me.

tjstienstra commented on doc/src/tutorials/physics/mechanics/rotating_disc_block.svg:
```diff
@@ -0,0 +1,55 @@
+    <g transform="matrix(0.9568503937007874 0.0 0.0 0.7775425369585419 806.0367026246719 158.5108997375328)"><clipPath id="p1.1">
+    <path d="m1.1368684E-13 0l100.0 0l0 40.95277l-100.0 0z" clip-rule="evenodd"/>
+    </clipPath>
+    <image clip-path="url(#p1.1)" fill="#000" width="100.0" height="43.0" x="0.0" y="0.0" preserveAspectRatio="none" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAAArCAYAAACO7C3tAAACS0lEQVR4Xu3ZN4gXQRgF8GcEA4IBMWClGEAEbQQLwUoQURTBTjEWgoUgWAja6IGhEiwEbQRRUKy0sJE/FmIhFmJhYS5EMGHAwvy+nZn7Zodbz2L339z7wePYeXNXzLJh9gARERERERERERERERkhFjKTy8FoBrOUmVgW0q5xzDHmBfOH+czszfrVzP3YWb4zB7JeWmRX ...
```
Can you clean up this SVG? It seems that it contains references to images. Personally, I like to use inkscape for this purpose.
And can you make it dark theme compatible refer to #24695

eh111eh replied:
I tried to delete `<image>` tags that have a reference (`xlink:href`), but it affects the images (some parts of the image are removed). I can redraw the diagram in Inkscape.

And I checked that the image displays correctly in dark mode.

<img width="770" alt="Screenshot 2024-08-12 at 19 05 45" src="https://github.com/user-attachments/assets/d9aa3093-f974-426c-9d73-30b10ea632f1">

eh111eh replied:
I installed Inkscape and redrew the schematic, but the arrows in the image disappear when I download it as SVG for some reason.

<img width="794" alt="Screenshot 2024-08-14 at 00 02 39" src="https://github.com/user-attachments/assets/3d2d1623-c326-49a7-ac08-9bf666d8a816">

<img width="576" alt="Screenshot 2024-08-14 at 00 07 04" src="https://github.com/user-attachments/assets/7470819d-688d-4c3c-883c-4ca8bd261741">

tjstienstra replied:
Nice top view, that makes the image/system easier to understand. I don't remember the exact reason that this problem may occur (I did see it before). An option is to try File-> clean document (I believe it was called something like that) and save it as plain svg. Another option is to just open one of the SVGs of a joint and copy an arrow from there and update the color.

eh111eh replied:
I directly edited the SVG file code, changing `fill-opacity:0` to `fill-opacity:1` and `stroke:context-stroke` to `stroke:#000000`. I also set the background color to white using `<rect width="100%" height="100%" fill="white"/>`. So now, it correctly displays in dark mode.

Peter230655 replied:
Small question: the coordinate systems A and N do not seem to line up for q_1 = 0, or so I understand the sketches. Any reason for this?
Thanks!

eh111eh replied:
Hi Peter, thanks for your interest in this PR! N is the inertia frame and A is the rotating frame, and frame A is used to define the position and velocity of q2 and q3. q1 won't be 0 at any time since I assume the initial velocities of q1, q2, and q3 are not zero, allowing us to use `CoulombKineticFriction` actuator, which is designed for kinetic friction.

Peter230655 replied:
What I meant was only this:
If you lign up N_x and A_x then N_y and A_y point in different directions. Was this intentional?
(N_z and A_z also have to point in opposite directions, else N or A is not right handed as per your sketch)

eh111eh replied:
![IMG_8726](https://github.com/user-attachments/assets/6b496495-b774-4868-b380-7f633fe806f0)

This is what I intend for N and A, so Nx and Ny don't line up with Ax and Ay, but Nz lines up with Az

eh111eh replied:
Sorry I drew a wrong picture for top view, this is the correct one
![IMG_8727](https://github.com/user-attachments/assets/d4acf731-3462-46db-99d1-2bc75bb014d0)

tjstienstra commented on doc/src/tutorials/physics/mechanics/rotating_disc_block_friction.rst:
```diff
@@ -0,0 +1,150 @@
+
+   >>> O.set_vel(N, 0)
+   >>> P.set_vel(N, r * u1 * A.y)
+   >>> Q.set_vel(N, P.vel(N) + v0 * A.x + v0 * A.y)
```
You don't require to add an initial velocity term to ensure a non-zero relative velocity of the block w.r.t the disc. The initial velocity should be specified using the numerical initial conditions when integrating.

tjstienstra commented on doc/src/tutorials/physics/mechanics/rotating_disc_block_friction.rst:
```diff
@@ -0,0 +1,150 @@
+   >>> normal_force = m * g
+   >>> pathway = SlidingPathway(P, Q, N)
+   >>> friction = me.CoulombKineticFriction(mu_k, normal_force, pathway, v_s=v_s, sigma=sigma, mu_s=mu_k)
+   >>> loads = friction.to_loads() + pathway.to_loads(force=None)
```
The `ForceActuator` uses the `Pathway.to_loads` method. You don't have to call it here.

eh111eh committed:
Fix sphinx-lint errors, Delete initial velocity of Q, Change pathway.to_loads to SlidingPathway.to_loads

eh111eh committed:
Set 3 reference frames, change velocities accordingly, change the force expression of `to_loads` method in `SlidingPathway`

Still have errors with kanes method

eh111eh committed:
Redraw schematic using Inkscape, set 2 reference frames, delete Q, rewrite force in `to_loads` as friction force, obtain kane's method

- Stribeck and viscous effects have not been added.
- Point Q has not been introduced.

tjstienstra commented on doc/src/tutorials/physics/mechanics/rotating_disc_block_friction.rst:
```diff
@@ -0,0 +1,160 @@
+==========================
+
+This page demonstrates how to use the functionalities in :obj:`sympy.physics.mechanics`
+and the **CoulombKineticFriction** actuator to model a sliding block on a rotating disc.
```
Can you use a `:class:` reference?

eh111eh replied:
Sure I changed `**CoulombKineticFriction**` to :class:`~.CoulombKineticFriction`.

eh111eh committed:
Use :class: reference to `CoulombKineticFriction`, update SVG file to display correctly in dark mode and show arrows, adjust SVG file size

eh111eh committed:
Change force in `to_loads` to introduce Stribeck and viscous terms in `CoulombKineticFriction`

eh111eh commented:
I tried to introduce the inertia of the disc and point _Q_ by
```python
import sympy as sm
import sympy.physics.mechanics as me
me.init_vprinting()

m, g, mu_k, mu_s, v_s, sigma, Ix, Iy, M = sm.symbols('m, g, mu_k, mu_s, v_s, sigma, Ix, Iy, M')
q1, q2, q3 = me.dynamicsymbols('q1:4')
q1d, q2d, q3d = me.dynamicsymbols('q1:4', level=1)
u1, u2, u3 = me.dynamicsymbols('u1:4')
u1d, u2d, u3d = me.dynamicsymbols('u1:4', level=1)

N = me.ReferenceFrame('N')
A = me.ReferenceFrame('A')
A.set_ang_vel(N, u1 * N.x)
#print(A.ang_vel_in(N))

O = me.Point('O')
P = O.locatenew('P', q2 * N.x + q3 * N.y)
Q = P.locatenew('Q', 0 * N.x + 0 * N.y)

O.set_vel(N, 0)
P.v2pt_theory(P, N, A)
Q.v2pt_theory(Q, N, A)
#print(P.v2pt_theory(P, N, A))
#print(Q.v2pt_theory(Q, N, A))

class SlidingPathway(me.PathwayBase):
   def __init__(self, fixed_point, contact_point, frame):
      """A custom pathway that moves along a rotating disc.

      Two points, ``fixed_point`` and ``contact_point``, start from the same location on the disc.
      ``contact_point``, the actual sliding point of the block, moves along the ``SlidingPathway``
      and is subjected to the friction force. ``SlidingPathway`` is valid only as long as
      the two points lie within the same plane, and it ensures the direction of the force is accurate.

      Parameters
      ==========
      fixed_point : Point
        Point fixed w.r.t the provided reference frame; the point on the disc.
      contact_point : Point
        Point with a time-dependent position w.r.t the fixed point, where the
        actuator forces will be applied. This point is assumed to be fixed to
        another body; the point on the block.
      frame : ReferenceFrame
        Reference frame in which the ``fixed_point`` has a zero velocity.

        """

      self.fixed_point = fixed_point
      self.contact_point = contact_point
      self.frame = frame

   @property
   def length(self):
    return self.contact_point.pos_from(self.fixed_point).magnitude()

   @property
   def extension_velocity(self):
       """Extension velocity of the pathway.

       The extension velocity of the pathway is the magnitude of the velocity of
       the ``contact_point`` relative to the frame in which the ``fixed_point``
       is stationary.

       """

       return self.contact_point.vel(self.frame).magnitude()

   def to_loads(self, force):
    """Loads in the correct format to be supplied to `KanesMethod`.

    Forces applied to the ``contact_point`` and ``fixed_point``
    based on the friction force.

    """

    direction = self.contact_point.vel(self.frame).normalize()
    force = self.extension_velocity * force

    return [
      me.Force(self.fixed_point, -force * direction),
      me.Force(self.contact_point, force * direction)
      ]

block = me.Particle('block', P, m)
disc = me.Particle('disc', Q, M)

inertia_disc = me.inertia(A, Ix, Iy, 0)
disc_body = me.RigidBody('disc_body', O, A, M, (inertia_disc, Q))

normal_force = m * g

pathway = SlidingPathway(P, Q, N)
friction = me.CoulombKineticFriction(mu_k, normal_force, pathway, v_s=v_s, sigma=sigma, mu_s=mu_s)

loads = friction.to_loads()
print("loads: ", loads)

BL = [block, disc]

kane = me.KanesMethod(
    N,
    q_ind=[q2, q3],
    u_ind=[u2, u3],
    kd_eqs=[q2d - u2, q3d - u3],
    bodies=BL
    )

fr, frstar = kane.kanes_equations(BL, loads)
eom = fr + frstar
print("eom: ", eom)
```
which results in
$$F_P = [\sigma * \sqrt((\dot{q_2})^2 + (\dot{q_3})^2) + (gm\mu_k + (gm\mu_s - gm\mu_k)*exp(-\frac{(\dot{q_2})^2 + (\dot{q_3})^2}{v_s^2}))*sgn(\sqrt((\dot{q_2})^2 + (\dot{q_3})^2))]*N.x + [\sigma * \sqrt((\dot{q_2})^2 + (\dot{q_3})^2) + (gm\mu_k + (gm\mu_s - gm\mu_k)*exp(-\frac{(\dot{q_2})^2 + (\dot{q_3})^2}{v_s^2}))*sgn(\sqrt((\dot{q_2})^2 + (\dot{q_3})^2))]*N.y$$

$$F_Q = -F_P$$

But the eom is Matrix([[-(M + m)*Derivative(u2(t), t)], [-(M + m)*Derivative(u3(t), t)]]).

I'm not sure if this is the right approach to introduce the disc.

eh111eh committed:
Include the inertia of the disc and its point Q, which starts from the same location as point P

Peter230655 commented:
Small point:
There is this line in the code:

direction = self.contact_point.vel(self.frame).normalize()

.normalize() gives numerical problems if the length of the vector to be normalized is very small.
If I cannot avoid it in my simulations, I do something like this:
length = sm.Max(vector.magnitude(), 1.e-15)
normalized_vector = vector / length

Of course not correct, but at least the simulation does not crash.

eh111eh committed:
Merge remote-tracking branch 'upstream/master' into coulombkineticfriction-example

eh111eh committed:
Replace Kane method with Lagrange method

eh111eh committed:
Resolve sphinx-lint error and one doctest error

eh111eh committed:
Update the diagram

Peter230655 reviewed (commented):
Should N_x and N_y not be perpenticular? Same question for A_x, A_y.

Peter230655 commented:
@eh111eh : I tried to implement the simplest friction model I could think of here:
https://www.pydy.org/pst-notebooks/examples/plot_block_on_disc_coulomb_friction.html#sphx-glr-examples-plot-block-on-disc-coulomb-friction-py
You know a lot about friction, I do not! If you have the time could you have a look at it and let me know the worst mistakes I made?
**Thanks!**
